# Ex3 - Getting and Knowing your Data

This time we are going to pull data directly from the internet.
Special thanks to: https://github.com/justmarkham for sharing the dataset and materials.

### Step 1. Import the necessary libraries

In [12]:
import numpy as np
import pandas as pd

### Step 2. Import the dataset from this [address](https://raw.githubusercontent.com/justmarkham/DAT8/master/data/u.user). 

In [13]:
url = 'https://raw.githubusercontent.com/justmarkham/DAT8/master/data/u.user'

url_data = pd.read_table(url, sep='|')
url_data.head(5)

user_id  age gender  occupation zip_code
0        1   24      M  technician    85711
1        2   53      F       other    94043
2        3   23      M      writer    32067
3        4   24      M  technician    43537
4        5   33      F       other    15213

### Step 3. Assign it to a variable called users and use the 'user_id' as index

In [14]:
users = url_data
users = users.set_index('user_id')

### Step 4. See the first 25 entries

In [15]:
users.head(25)

age gender     occupation zip_code
user_id                                    
1         24      M     technician    85711
2         53      F          other    94043
3         23      M         writer    32067
4         24      M     technician    43537
5         33      F          other    15213
6         42      M      executive    98101
7         57      M  administrator    91344
8         36      M  administrator    05201
9         29      M        student    01002
10        53      M         lawyer    90703
11        39      F          other    30329
12        28      F          other    06405
13        47      M       educator    29206
14        45      M      scientist    55106
15        49      F       educator    97301
16        21      M  entertainment    10309
17        30      M     programmer    06355
18        35      F          other    37212
19        40      M      librarian    02138
20        42      F      homemaker    95660
21        26      M         writer    30068
22        25      M         writer    40206
23        30      F         artist    48197
24        21      F         artist    94533
25        39      M       engineer    55107

In [16]:
users.iloc[:25, :]

age gender     occupation zip_code
user_id                                    
1         24      M     technician    85711
2         53      F          other    94043
3         23      M         writer    32067
4         24      M     technician    43537
5         33      F          other    15213
6         42      M      executive    98101
7         57      M  administrator    91344
8         36      M  administrator    05201
9         29      M        student    01002
10        53      M         lawyer    90703
11        39      F          other    30329
12        28      F          other    06405
13        47      M       educator    29206
14        45      M      scientist    55106
15        49      F       educator    97301
16        21      M  entertainment    10309
17        30      M     programmer    06355
18        35      F          other    37212
19        40      M      librarian    02138
20        42      F      homemaker    95660
21        26      M         writer    30068
22        25      M         writer    40206
23        30      F         artist    48197
24        21      F         artist    94533
25        39      M       engineer    55107

### Step 5. See the last 10 entries

In [17]:
users.tail(10)

age gender     occupation zip_code
user_id                                    
934       61      M       engineer    22902
935       42      M         doctor    66221
936       24      M          other    32789
937       48      M       educator    98072
938       38      F     technician    55038
939       26      F        student    33319
940       32      M  administrator    02215
941       20      M        student    97229
942       48      F      librarian    78209
943       22      M        student    77841

### Step 6. What is the number of observations in the dataset?

In [18]:
len(users)

943

### Step 7. What is the number of columns in the dataset?

In [19]:
len(users.columns)

4

### Step 8. Print the name of all the columns.

In [20]:
for col in users.columns:
    print(col)

age
gender
occupation
zip_code


### Step 9. How is the dataset indexed?
It's an Int64 Index which runs from 1 to 943, therefore with a 943 length, and named under 'user_id'

In [21]:
users.index

Int64Index([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,
            ...
            934, 935, 936, 937, 938, 939, 940, 941, 942, 943],
           dtype='int64', name='user_id', length=943)

### Step 10. What is the data type of each column?

In [22]:
users.dtypes

age            int64
gender        object
occupation    object
zip_code      object
dtype: object

### Step 11. Print only the occupation column

In [23]:
users['occupation']

user_id
1         technician
2              other
3             writer
4         technician
5              other
           ...      
939          student
940    administrator
941          student
942        librarian
943          student
Name: occupation, Length: 943, dtype: object

### Step 12. How many different occupations are in this dataset?

In [24]:
users['occupation'].nunique()

21

In [25]:
users['occupation'].value_counts().count()

21

In [26]:
len(users.groupby('occupation'))

21

### Step 13. What is the most frequent occupation?

In [27]:
users['occupation'].value_counts()
most_frequent_occupation = users['occupation'].value_counts()
most_frequent_occupation.head(1)

student    196
Name: occupation, dtype: int64

In [28]:
users.occupation.mode()

0    student
dtype: object

In [42]:
users_ = users.groupby('occupation').count()
users__ = users_.sort_values('age', ascending=False)
users__

age  gender  zip_code
occupation                          
student        196     196       196
other          105     105       105
educator        95      95        95
administrator   79      79        79
engineer        67      67        67
programmer      66      66        66
librarian       51      51        51
writer          45      45        45
executive       32      32        32
scientist       31      31        31
artist          28      28        28
technician      27      27        27
marketing       26      26        26
entertainment   18      18        18
healthcare      16      16        16
retired         14      14        14
lawyer          12      12        12
salesman        12      12        12
none             9       9         9
homemaker        7       7         7
doctor           7       7         7

### Step 14. Summarize the DataFrame.

In [23]:
users.describe()

age
count  943.000000
mean    34.051962
std     12.192740
min      7.000000
25%     25.000000
50%     31.000000
75%     43.000000
max     73.000000

### Step 15. Summarize all the columns

In [41]:
users.describe(include='all')

age gender occupation zip_code
count   943.000000    943        943      943
unique         NaN      2         21      795
top            NaN      M    student    55414
freq           NaN    670        196        9
mean     34.051962    NaN        NaN      NaN
std      12.192740    NaN        NaN      NaN
min       7.000000    NaN        NaN      NaN
25%      25.000000    NaN        NaN      NaN
50%      31.000000    NaN        NaN      NaN
75%      43.000000    NaN        NaN      NaN
max      73.000000    NaN        NaN      NaN

### Step 16. Summarize only the occupation column

In [30]:
users['occupation'].describe()

count         943
unique         21
top       student
freq          196
Name: occupation, dtype: object

### Step 17. What is the mean age of users?

In [48]:
users['age'].mean()

34.05196182396607

In [31]:
round(users.age.mean())

34

### Step 18. What is the age with least occurrence?

In [37]:
least_occurence_age = users['age'].value_counts()
least_occurence_age.tail(5)


11    1
10    1
73    1
66    1
7     1
Name: age, dtype: int64

In [36]:
users['age'].value_counts().tail()

11    1
10    1
73    1
66    1
7     1
Name: age, dtype: int64